### ISSUE TITLE
CUDA AveragePool fails with CUDNN_STATUS_NOT_SUPPORTED when batch dimension reaches 65536

### ISSUE THREAD

**Issue by Alcoholrithm** [microsoft/onnxruntime#29819]
### Description

  `AveragePool` executed by `CUDAExecutionProvider` fails with
  `CUDNN_STATUS_NOT_SUPPORTED` when the effective batch dimension reaches 65536.

  The same operation succeeds:

  - With `N=65535`
  - In PyTorch CUDA with `N=196608`
  - After replacing `AveragePool` with `Reshape + ReduceMean`

  This appears to be specific to ONNX Runtime's cuDNN pooling path.

  ### Environment

  - OS: Ubuntu 24.04
  - GPU: NVIDIA GeForce RTX 5090
  - NVIDIA driver: 575.64.03
  - ONNX Runtime: 1.23.2
  - CUDA: 12.9
  - cuDNN: 9.10.2
  - Python: 3.10
  - ONNX opset: 18

  The problem was also reproduced with newer ONNX Runtime/cuDNN combinations
  and NVIDIA driver 610.43.02.

  ### Expected behavior

  AveragePool should execute for this valid tensor shape, or ONNX Runtime
  should provide a non-cuDNN CUDA fallback when cudnnPoolingForward returns
  CUDNN_STATUS_NOT_SUPPORTED.

### Minimal reproduction

  ```python
  import numpy as np
  import onnx
  import onnxruntime as ort
  from onnx import TensorProto, helper


  def make_model(path, n):
      input_shape = [n, 1, 1, 512]
      output_shape = [n, 1, 1, 256]

      graph = helper.make_graph(
          [
              helper.make_node(
                  "AveragePool",
                  ["input"],
                  ["output"],
                  name="node_avg_pool2d",
                  kernel_shape=[1, 2],
                  strides=[1, 2],
                  pads=[0, 0, 0, 0],
                  ceil_mode=0,
                  count_include_pad=1,
              )
          ],
          "large_batch_average_pool",
          [
              helper.make_tensor_value_info(
                  "input", TensorProto.FLOAT, input_shape
              )
          ],
          [
              helper.make_tensor_value_info(
                  "output", TensorProto.FLOAT, output_shape
              )
          ],
      )

      model = helper.make_model(
          graph,
          opset_imports=[helper.make_opsetid("", 18)],
          ir_version=10,
      )
      onnx.checker.check_model(model)
      onnx.save(model, path)


  for n in [65535, 65536]:
      path = f"avgpool_{n}.onnx"
      make_model(path, n)

      session = ort.InferenceSession(
          path,
          providers=["CUDAExecutionProvider"],
      )
      x = np.zeros((n, 1, 1, 512), dtype=np.float32)

      try:
          y = session.run(None, {"input": x})
          print(n, "SUCCESS", y[0].shape)
      except Exception as exc:
          print(n, "FAILED", exc)
```
  Observed result:

  65535 SUCCESS (65535, 1, 1, 256)
  65536 FAILED CUDNN_STATUS_NOT_SUPPORTED

  ### Error

  CUDNN failure 3000: CUDNN_STATUS_NOT_SUPPORTED
  file=/onnxruntime_src/onnxruntime/core/providers/cuda/nn/pool.cc
  line=240
  expr=PoolingForwardHelper(
      GetCudnnHandle(context),
      pooling_desc,
      &alpha,
      x_tensor,
      x_data,
      &beta,
      y_tensor,
      y_data
  )

  ### Comparison with PyTorch

  The equivalent PyTorch operation succeeds on the same GPU:

```python
  import torch
  import torch.nn.functional as F

  x = torch.zeros((196608, 1, 1, 512), device="cuda")
  y = F.avg_pool2d(x, kernel_size=(1, 2), stride=(1, 2))
  print(y.shape)
```
  Replacing the ONNX node with the following graph also succeeds using only
  CUDAExecutionProvider:

  Reshape [N, C, H, W] -> [N, C, H, W/2, 2]
  ReduceMean(axis=-1)

  The replacement produced bit-exact outputs in the complete model.

### Urgency

There is no fixed external deadline, but this is blocking validation and deployment of the high-resolution model.

### Platform

Linux

### OS Version

Ubuntu 24.04

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.23.2

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

CUDA

**Comment by Saad-Mallebhari** [comment]
Hi @Alcoholrithm

Thanks for the detailed issue report and the exact reproduction snippet.

Root Cause

In `onnxruntime/core/providers/cuda/nn/pool.cc` (line 240), `PoolOp::ComputeInternal` dispatches to `cudnnPoolingForward`. Because cuDNN's legacy pooling descriptor API enforces a 16-bit integer limit ($65,535$) on spatial and batch dimension extents, `cudnnPoolingForward` returns `CUDNN_STATUS_NOT_SUPPORTED` when batch $N \ge 65536$. ONNX Runtime's CUDA provider currently lacks a fallback path when cuDNN returns `CUDNN_STATUS_NOT_SUPPORTED`, causing valid large-batch inference graphs to fail.

Proposed Solution

1. In `onnxruntime/core/providers/cuda/nn/pool.cc`, catch `CUDNN_STATUS_NOT_SUPPORTED` from `cudnnPoolingForward` (or pre-check if batch $N > 65535$).
2. Implement a fallback mechanism in `PoolOp` that dispatches to ONNX Runtime's native CUDA pooling kernel when cuDNN rejects large dimension extents.
3. Add a unit test in `onnxruntime/test/providers/cuda/cuda_provider_test.cc` asserting that `AveragePool` executes without errors for batch sizes $N \ge 65536$.

Would you like me to open a Pull Request with this cuDNN pooling fallback fix and accompanying unit tests?
